File: 01-content-factory/python/agents/social_agent.py
"""
Social Media Agent
Creates LinkedIn carousels, X/Twitter threads, and Instagram posts with image prompts
"""
from typing import List
from ..models import BrandVoice, ContentTopic, BlogPost, SocialPost, ContentType
from ..providers.claude import ClaudeProvider
from ..utils import safe_extract_json


class SocialAgent:
    """
    Expert social media content creator that generates platform-native content
    including LinkedIn carousels, X threads, and Instagram posts with Midjourney prompts.
    """
    
    SYSTEM_PROMPT = """You are a viral social media content expert who creates platform-native posts that drive massive engagement.

You understand:
- LinkedIn: Professional tone, thought leadership, carousel best practices
- Twitter/X: Thread hooks, punchy sentences, engagement tactics
- Instagram: Visual storytelling, caption hooks, hashtag strategy

Each post must feel native to its platform, not cross-posted.
Include Midjourney image prompts that will create scroll-stopping visuals."""

    LINKEDIN_PROMPT = """Create 8 LinkedIn carousel posts from this blog content:

**Blog Title:** {title}
**Blog Content:** {content}
**Brand Voice:** {brand_voice}
**Target Audience:** {target_audience}

For EACH post, provide:
1. Hook (first line that stops the scroll)
2. Carousel slides (5-8 slides with key points)
3. Full caption with CTA
4. Midjourney prompt for the cover slide

Return JSON array:
[
  {{
    "id": "linkedin_1",
    "hook": "First attention-grabbing line",
    "carousel_slides": ["Slide 1 text", "Slide 2 text", ...],
    "content": "Full caption with hashtags",
    "hashtags": ["#tag1", "#tag2"],
    "media_prompt": "Midjourney prompt for carousel cover: professional, modern..."
  }}
]"""

    TWITTER_PROMPT = """Create 4 X/Twitter threads from this blog content:

**Blog Title:** {title}
**Blog Content:** {content}
**Brand Voice:** {brand_voice}

For EACH thread, provide:
1. Hook tweet (must be provocative/valuable enough to make people click)
2. Thread content (8-12 tweets)
3. CTA tweet at the end

Return JSON array:
[
  {{
    "id": "twitter_1",
    "hook": "Opening hook tweet",
    "content": "Full thread with each tweet separated by \\n\\n",
    "hashtags": ["#tag1"]
  }}
]"""

    INSTAGRAM_PROMPT = """Create 4 Instagram posts from this blog content:

**Blog Title:** {title}
**Blog Content:** {content}
**Brand Voice:** {brand_voice}
**Target Audience:** {target_audience}

For EACH post, provide:
1. Caption with hook, value, and CTA
2. Hashtags (20-30 relevant hashtags)
3. Midjourney prompt for the visual

Return JSON array:
[
  {{
    "id": "instagram_1",
    "content": "Full caption with line breaks",
    "hashtags": ["#tag1", "#tag2", ...],
    "media_prompt": "Midjourney prompt: aesthetic, Instagram-worthy..."
  }}
]"""

    def __init__(self):
        self.llm = ClaudeProvider()
    
    async def generate_from_blog(
        self,
        blog: BlogPost,
        brand_voice: BrandVoice,
        platforms: List[ContentType] = None,
    ) -> List[SocialPost]:
        """Generate social posts from a blog post"""
        
        if platforms is None:
            platforms = [ContentType.LINKEDIN, ContentType.TWITTER, ContentType.INSTAGRAM]
        
        all_posts = []
        
        if ContentType.LINKEDIN in platforms:
            linkedin_posts = await self._generate_linkedin(blog, brand_voice)
            all_posts.extend(linkedin_posts)
        
        if ContentType.TWITTER in platforms:
            twitter_posts = await self._generate_twitter(blog, brand_voice)
            all_posts.extend(twitter_posts)
        
        if ContentType.INSTAGRAM in platforms:
            instagram_posts = await self._generate_instagram(blog, brand_voice)
            all_posts.extend(instagram_posts)
        
        return all_posts
    
    async def _generate_linkedin(
        self,
        blog: BlogPost,
        brand_voice: BrandVoice,
    ) -> List[SocialPost]:
        """Generate LinkedIn carousel posts"""
        
        user = self.LINKEDIN_PROMPT.format(
            title=blog.title,
            content=blog.content[:3000],
            brand_voice=brand_voice.tone,
            target_audience=brand_voice.target_audience,
        )
        
        response = await self.llm.generate_json(self.SYSTEM_PROMPT, user)
        posts_data = safe_extract_json(response, expect_array=True)
        
        posts = []
        for i, post in enumerate(posts_data):
            posts.append(SocialPost(
                id=post.get("id", f"linkedin_{blog.topic_id}_{i+1}"),
                topic_id=blog.topic_id,
                platform="linkedin",
                content=post["content"],
                hashtags=post.get("hashtags", []),
                media_prompt=post.get("media_prompt"),
                carousel_slides=post.get("carousel_slides"),
            ))
        
        return posts
    
    async def _generate_twitter(
        self,
        blog: BlogPost,
        brand_voice: BrandVoice,
    ) -> List[SocialPost]:
        """Generate Twitter/X threads"""
        
        user = self.TWITTER_PROMPT.format(
            title=blog.title,
            content=blog.content[:3000],
            brand_voice=brand_voice.tone,
        )
        
        response = await self.llm.generate_json(self.SYSTEM_PROMPT, user)
        posts_data = safe_extract_json(response, expect_array=True)
        
        posts = []
        for i, post in enumerate(posts_data):
            posts.append(SocialPost(
                id=post.get("id", f"twitter_{blog.topic_id}_{i+1}"),
                topic_id=blog.topic_id,
                platform="twitter",
                content=post["content"],
                hashtags=post.get("hashtags", []),
            ))
        
        return posts
    
    async def _generate_instagram(
        self,
        blog: BlogPost,
        brand_voice: BrandVoice,
    ) -> List[SocialPost]:
        """Generate Instagram posts"""
        
        user = self.INSTAGRAM_PROMPT.format(
            title=blog.title,
            content=blog.content[:3000],
            brand_voice=brand_voice.tone,
            target_audience=brand_voice.target_audience,
        )
        
        response = await self.llm.generate_json(self.SYSTEM_PROMPT, user)
        posts_data = safe_extract_json(response, expect_array=True)
        
        posts = []
        for i, post in enumerate(posts_data):
            posts.append(SocialPost(
                id=post.get("id", f"instagram_{blog.topic_id}_{i+1}"),
                topic_id=blog.topic_id,
                platform="instagram",
                content=post["content"],
                hashtags=post.get("hashtags", []),
                media_prompt=post.get("media_prompt"),
            ))
        
        return posts
    
    async def generate_from_topic(
        self,
        topic: ContentTopic,
        brand_voice: BrandVoice,
    ) -> List[SocialPost]:
        """Generate social posts directly from a topic (without blog)"""
        
        prompt = f"""Create social media posts for this topic:

**Topic:** {topic.title}
**Angle:** {topic.angle}
**Keywords:** {", ".join(topic.keywords)}
**Brand Voice:** {brand_voice.tone}
**Target Audience:** {brand_voice.target_audience}

Create:
- 2 LinkedIn posts with carousel slides
- 2 Twitter/X threads
- 2 Instagram posts

Return JSON array with platform, content, hashtags, media_prompt, carousel_slides fields."""

        response = await self.llm.generate_json(self.SYSTEM_PROMPT, prompt)
        posts_data = json.loads(self._extract_json(response))
        
        posts = []
        for i, post in enumerate(posts_data):
            posts.append(SocialPost(
                id=f"social_{topic.id}_{i+1}",
                topic_id=topic.id,
                platform=post.get("platform", "linkedin"),
                content=post["content"],
                hashtags=post.get("hashtags", []),
                media_prompt=post.get("media_prompt"),
                carousel_slides=post.get("carousel_slides"),
            ))
        
        return posts

File: 01-content-factory/agents/social-agent.ts
import { generateWithClaude } from "../integrations/anthropic";
import { generateSocialMediaGraphic } from "../integrations/nano-banana-pro";
import type { ClientBrief, ContentTopic, GeneratedContent, ContentType, AgentResponse } from "../types";

const PLATFORM_CONFIGS = {
  linkedin: {
    maxChars: 3000,
    style: "Professional, thought-leadership focused, use line breaks for readability, include 3-5 relevant hashtags",
    tone: "authoritative yet approachable",
  },
  twitter: {
    maxChars: 280,
    style: "Punchy, conversational, use emojis sparingly, include 1-2 hashtags",
    tone: "witty and engaging",
  },
  instagram: {
    maxChars: 2200,
    style: "Visual-first caption, storytelling format, heavy hashtag usage (20-30), include call-to-action",
    tone: "authentic and relatable",
  },
};

const SYSTEM_PROMPT = `You are a social media content expert who creates scroll-stopping posts for various platforms. You understand:

- Platform-specific best practices and algorithms
- Engagement-driving hooks and formats
- Hashtag strategies for discoverability
- Call-to-action optimization
- Visual content suggestions

Create content that drives engagement, shares, and conversions.`;

export async function generateSocialPost(
  topic: ContentTopic,
  brief: ClientBrief,
  platform: 'linkedin' | 'twitter' | 'instagram'
): Promise<AgentResponse<GeneratedContent>> {
  try {
    const config = PLATFORM_CONFIGS[platform];
    
    const userPrompt = `Create a ${platform.toUpperCase()} post for ${brief.clientName}.

**Topic:** ${topic.title}
**Angle:** ${topic.angle}
**Keywords:** ${topic.keywords.join(', ')}
**Brand Voice:** ${brief.brandVoice}
**Target Audience:** ${brief.targetAudience}

**Platform Requirements:**
- Max characters: ${config.maxChars}
- Style: ${config.style}
- Tone: ${config.tone}

Create a highly engaging post that:
1. Opens with a hook that stops scrolling
2. Delivers value or insight
3. Ends with engagement prompt or CTA
4. Includes appropriate hashtags

Output ONLY the post content, ready to publish.`;

    const content = await generateWithClaude(SYSTEM_PROMPT, userPrompt, {
      maxTokens: 1024,
      temperature: 0.8,
    });

    const hashtagMatch = content.match(/#\w+/g);
    
    let imageDataUrl: string | undefined;
    
    const geminiKey = process.env.GEMINI_API_KEY || process.env.AI_INTEGRATIONS_GEMINI_API_KEY;
    if (geminiKey) {
      try {
        console.log(`[SocialAgent] Generating image with Nano Banana Pro for ${platform} post...`);
        const imageResult = await generateSocialMediaGraphic(
          topic.title,
          platform,
          brief.brandVoice
        );
        
        if (imageResult.success && imageResult.imageDataUrl) {
          imageDataUrl = imageResult.imageDataUrl;
          console.log(`[SocialAgent] Image generated for ${platform} post`);
        } else {
          console.log(`[SocialAgent] Image generation failed for ${platform}: ${imageResult.error}`);
        }
      } catch (imageError: any) {
        console.log(`[SocialAgent] Image generation error for ${platform}:`, imageError.message);
      }
    } else {
      console.log(`[SocialAgent] Skipping image generation - no Gemini API key configured`);
    }
    
    const generatedContent: GeneratedContent = {
      id: `${platform}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      topicId: topic.id,
      clientId: brief.clientId,
      type: platform as ContentType,
      title: topic.title,
      content: content.trim(),
      metadata: {
        characterCount: content.length,
        hashtags: hashtagMatch || [],
        imageDataUrl,
      },
      status: 'draft',
      createdAt: new Date(),
    };

    return {
      success: true,
      data: generatedContent,
    };
  } catch (error: any) {
    return {
      success: false,
      error: error.message || `Failed to generate ${platform} post`,
    };
  }
}

export async function generateAllSocialPosts(
  topic: ContentTopic,
  brief: ClientBrief,
  platforms: ('linkedin' | 'twitter' | 'instagram')[] = ['linkedin', 'twitter', 'instagram']
): Promise<AgentResponse<GeneratedContent[]>> {
  try {
    const results = await Promise.all(
      platforms.map(platform => generateSocialPost(topic, brief, platform))
    );

    const successfulPosts = results
      .filter(r => r.success && r.data)
      .map(r => r.data as GeneratedContent);

    return {
      success: successfulPosts.length > 0,
      data: successfulPosts,
    };
  } catch (error: any) {
    return {
      success: false,
      error: error.message || "Failed to generate social posts",
    };
  }
}
